### ISSUE TITLE
merge dask and block-by-block IO in ifgram_inversion.py

### ISSUE THREAD

**Issue by yunjunz** [insarlab/MintPy#347]
In the current implementation, we achieve memory efficiency via block-by-block reading/writing and achieve computing efficiency via dask, separately. However, we should be able to apply both of them, by applying dask job submission within each block. This won't require HDF5 parallel writing support. A computing and memory efficient ifgram_inversion.py is relatively urgent.

More specifically, we have a primary `box_list` to save memory ([L1008](https://github.com/insarlab/MintPy/blob/master/mintpy/ifgram_inversion.py#L1008)). Dask part of the code is further splitting this box_list into a list of boxes with a smaller size ([L1121](https://github.com/insarlab/MintPy/blob/master/mintpy/ifgram_inversion.py#L1121)) and submitting jobs for the entire stack. This procedure can be changed as: 

+ for each box element in the primary `box_list`, if parallel is turned on, split the box into a secondary list of boxes at [here](https://github.com/insarlab/MintPy/blob/master/mintpy/ifgram_inversion.py#L1057)
+ submit jobs for this secondary list of boxes
+ when all of them are finished, write this block to disk/file as [here](https://github.com/insarlab/MintPy/blob/master/mintpy/ifgram_inversion.py#L1072)
+ repeat this procedure for the next box in the primary box_list.

- [x] Could you please try it @Ovec8hkin ?

Below are a few minor suggestions related with ifgram_inversion.py, with priority from high to low. We could check them after the above works.

- [x] 1. the current printout message in the dask part is not very intuitive to me, in the sense that I don't know what's the current status of the processing with dask, in a progressive way. Could you clean it up?

- [x] 2. Since the procedure above is submitting jobs for each block, we might want fewer blocks splitter primarily, maybe from the current chunk_size of 1e8 (~1G in memory) to 4e8 (~4G in memory). And add a template option, i.e. `mintpy.networkInversion.chunkSize`, for it so that one could change according to their resource.

- [x] 3. Update [document](https://mintpy.readthedocs.io/en/latest/dask/) for the newly added `local` cluster. The local cluster is the only available option to me currently, (I believe) same as the majority of users. So it is important. 

- [ ] 4. Check if it's possible to copy the `mintpy.yaml` file from source code to the local working directly, and use it to feed dask, instead of using the one from `.config/dask/*`, as we mentioned at #343.

- [x] 5. `data`at [L1128](https://github.com/insarlab/MintPy/blob/master/mintpy/ifgram_inversion.py#L1128) can be simplified as `data = (subbox, inps)` because all the other parameters besides subbox are fixed and are the same for different subboxes.

- [x] 6. Regarding the document, it will be very cool to have a performance figure showing how much computation time we spent with the different number of cores we used (2, 4, 8, 16, 32, 64, etc.). This is useful for benchmarking as well.

**Comment by Ovec8hkin** [comment]
Pertaining to the `ifgram_inversion.py` suggestions:

1. I assume you are referring to the fact that print statements appear to show up in a random order, or are repeated various number of times. This is due to the fact that each core is running in parallel, at slightly different speeds, which leads to the same message being printed by each core. I agree it is unwieldy, if you have suggestions as to how to fix this, I would be glad to try some out.

2. I was actually going to inquire about this, because the Kilauea dataset splits into something like 580 blocks, and takes forever to process. I would be interested in messing with this chunk size, manually at first, to see if that improves performance on large datasets at all. If I find that it does, I definitely agree with making a template option for it.

3. I will make this update soon.

4. I will look through the dask documentation and inquire on the dask git repo if this is possible. If it is not, it would be fairly simple, though tedious, to write a wrapper to read the dask options from a local template file and pass them directly. I actually think that might be a good way to go, since Falk pointed out that we can change dask options for each dataset at the moment, which could be very important.

5. I can do this, but it is very low priority, as you said.

6. I will run some test in the background on Stampede tomorrow for 1, 2, 4, 8, 16, 32, and 64 nodes for the Fernandina and Galapagos datasets (that's what I have on Stampede at the moment), and create a figure with the results. Once I have the Kilauea dataset working I can add that data as well.

**Comment by Ovec8hkin** [comment]
As to the block-by-block writing, I can work on this, but I want to make sure the local cluster stuff is working to everybody's liking and specifications first. So I imagine I won't get to it until next week at the very earliest and likely not until the following week. I will keep you updated.

**Comment by falkamelung** [comment]
Re 4, I agree. Wrappers are better than environment variables. We should not have any system-specific  options in the *cfg file.  There could be a local, system-specific mintpy_dask.yaml.

**Comment by Ovec8hkin** [comment]
> More specifically, we have a primary box_list to save memory (L1008). Dask part of the code is further splitting this box_list into a list of boxes with a smaller size (L1121) and submitting jobs for the entire stack. This procedure can be changed as
> 1. for each box element in the primary box_list, if parallel is turned on, split the box into a secondary list of boxes at **here**
> 2. submit jobs for this secondary list of boxes
> 3. when all of them are finished, write this block to disk/file as **here**
> 4. repeat this procedure for the next box in the primary box_list.

Regarding step 1: the location you linked to is already nested within a the condition `if not inps.parallel`, thus we can't rely on a check to `inps.parallel` to enable the block-by-block writing as described above, since it will always by False

Additionally, I don't quite understand what doing the block-by-block writing in Dask at the location described above is saving computational as compared to what you ave implemented right now. If you want the `ifgram_inversion` to go faster just decrease the chunk size to make more blocks.

**Comment by yunjunz** [comment]
3 is completed by #349

**Comment by yunjunz** [comment]
The main task and 1, 2, 5 are completed by #351.

**Comment by yunjunz** [comment]
@2gotgrossman do you have any suggestions regarding using local YAML file for dask configuration?

**Comment by 2gotgrossman** [comment]
@yunjunz, based on the [Dask Jobqueue documentation](https://jobqueue.dask.org/en/latest/configuration-setup.html#managing-configuration-files), yaml config files should live in `~/.config/dask` or `/etc/dask`. 

However, looking at the Dask code, it looks like you may be able to configure an additional location by setting `DASK_CONFIG`. This is risky as it is not documented in `dask_jobqueue` (although it is documented [here](https://docs.dask.org/en/latest/configuration.html) in `dask`) so this may change. https://github.com/dask/dask/blob/c38c360dd51fc01d6d15bd2636adf906622e44c0/dask/config.py#L18-L29

I would recommend keeping the Dask `.yaml` files separate from other `.yaml` files, since we want to keep the namespaces separate (if a key is used by us and also by Dask for 2 separate purposes, weird bugs might crop up if we expect that key to only be used internally to MintPy)

**Comment by yunjunz** [comment]
Thank you @2gotgrossman. This is similar to what we know now. @falkamelung I guess one could setup several `config`s in a YAML file and switch it using the option in the template file.

**Comment by Ovec8hkin** [comment]
@2gotgrossman I think what @falkamelung desires is placing a dask.yaml file in each project directory so that the dask configuration for each project can be different and stays local to the project. Personally, this sounds like not a bad idea, though it would require us to write a wrapper  to read the local `dask.yaml` file and pass the info to the dask cluster manually. Thoughts?

**Comment by yunjunz** [comment]
Hi @Ovec8hkin, it seems like in our [dask doc](https://mintpy.readthedocs.io/en/latest/dask/), the detailed description for the "processes" and "walltime" are not up-to-date with the code. Could you check and update them if so?

**Comment by Ovec8hkin** [comment]
What seems to be the problem with them? Both of those parameters are used when running dask via am HPC cluster, which has been modified in any way for a few weeks.

**Comment by yunjunz** [comment]
@Ovec8hkin You are right. There is no issue, my misjudge.

**Comment by yunjunz** [comment]
Among the objectives of this issue, all have been implemented and merged, except for the potential local YAML file. Since there is no practical way of implementing the local YAML file from the discussions above, I will close this issue. The purpose of this functionality can be achieved via `mintpy.compute.config` already.
